src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/ExcusalRefusalLetterListRepositoryImpl.java

```diff
@@ -1,79 +1,193 @@
+                .isActive(tuple.get(QJurorPool.jurorPool.isActive))
+                .build())
+            .collect(Collectors.toList());
+
+        // for each juror select the most recent excusal refused date and remove the other entries
+        ConcurrentHashMap<String, ExcusalRefusedLetterList> excusalRefusedLetterListMap = new ConcurrentHashMap<>();
+        excusalRefusedLetterLists.forEach(excusalRefusedLetterList -> {
+            String jurorNumber = excusalRefusedLetterList.getJurorNumber();
+            if (excusalRefusedLetterListMap.containsKey(jurorNumber)) {
+                ExcusalRefusedLetterList existingExcusalRefusedLetterList =
+                    excusalRefusedLetterListMap.get(jurorNumber);
+                if (excusalRefusedLetterList.getDateExcused()
+                    .isAfter(existingExcusalRefusedLetterList.getDateExcused())) {
+                    excusalRefusedLetterListMap.put(jurorNumber, excusalRefusedLetterList);
+                }
+            } else {
+                excusalRefusedLetterListMap.put(jurorNumber, excusalRefusedLetterList);
+            }
+        });
+
+        List<String> poolNumbers =
+            excusalRefusedLetterLists.stream().map(ExcusalRefusedLetterList::getPoolNumber).distinct().toList();
+
+        // run query to get the most recent excusal refused date for each juror
+        List<Tuple> printedDates = getDatePrinted(owner, poolNumbers);
+
+        // update the excusal refused letter printed date for each juror
+        printedDates.forEach(tuple -> {
+            String jurorNumber = tuple.get(QJuror.juror.jurorNumber);
+            if (excusalRefusedLetterListMap.containsKey(jurorNumber)) {
+                ExcusalRefusedLetterList excusalRefusedLetterList = excusalRefusedLetterListMap.get(jurorNumber);
+                LocalDateTime datePrinted = tuple.get(QJurorHistory.jurorHistory.dateCreated);
+                LocalDate datePrintedDateOnly = datePrinted.toLocalDate();
+                if (datePrintedDateOnly.equals(excusalRefusedLetterList.getDateExcused())
+                    || datePrintedDateOnly.isAfter(excusalRefusedLetterList.getDateExcused())) {
+                    if (!searchCriteria.includePrinted()) {
+                        excusalRefusedLetterListMap.remove(jurorNumber);
+                    } else {
+                        excusalRefusedLetterList.setDatePrinted(datePrinted);
+                    }
+                }
+            }
+        });
+
+        Comparator<ExcusalRefusedLetterList> datePrintComparator = (o1, o2) -> {
+            if (o1.getDatePrinted() == null && o2.getDatePrinted() == null) {
+                return 0;
+            } else if (o1.getDatePrinted() == null) {
+                return -1;
+            } else if (o2.getDatePrinted() == null) {
+                return 1;
+            } else {
+                return o2.getDatePrinted().compareTo(o1.getDatePrinted());
+            }
+        };
```

Comment:
Could this be a final static at the top of the class?
To save instantiating it every call

Reply:
updated